doc/design/logging.md

```diff
@@ -0,0 +1,385 @@
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independently but
+in combination with the error handler.
+
+## Design
+
+### Considerations
+
+1. Similar to other iceoryx constructs, heap usage, exceptions and recursive
+   calls are forbidden. This also means that `std::string` and
+   `std::stringstream` cannot be used.
+
+2. Since the logger shall be used by the `iceoryx_hoofs` itself, the base
+   implementation can also not make use of functionality in e.g. `iox::cxx` or
+   `iox::posix`
+
+3. To have a minimal runtime overhead, string formatting shall only happen when
+   the log level is above the output threshold. This lazy evaluation shall also
+   apply to other potentially expensive operations like the execution of a
+   function call.
+
+4. Additionally, the minimal log level shall be configurable by a compile time
+   switch and everything below this log level should be compiled to no-ops. With
+   this, developer can make excessive use of the most verbose log level and
+   disable it by default. When debugging, this can be turned on and help to find
+   the root of the problem.
+
+5. There shall also be the possibility to alter the behavior of the logger via
+   environment variables. This can be functionality like setting the log level or
+   deciding whether the timestamps or file names shall be part of a log message.
+
+6. Since iceoryx is often used as transport layer in a higher level framework, it
+   shall be possible to replace the default logger at runtime and delegate the
+   log messages to the logger of such a framework. To prevent misuse, this can
+   only be done during the initialization phase. Afterwards an error message shall
+   be logged in the new and old logger and the error handler shall be called with
+   a moderate error level.
+
+7. The default logger shall be replaceable via the platform abstraction in order
+   to use the dedicated logger for log messages which are emitted before it could
+   be replaced at runtime, e.g. from global objects before `main`.
+
+8. Additionally, there shall be a compile time option to forward all log messages
+   independent of the log level threshold for cases where the logger framework
+   does the filtering itself.
+
+9. For tests, the default logger shall be replaced to suppress the log messages in
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally, there shall be an
+   environment variable to circumvent the suppression when desired.
+
```

Comment:
Since we already encountered the mutex in the logger. Should we add a point were we state that the logger shall be usable/exchangeable without the usage of a mutex.
Here I even would go so far to restrict it to an `atomic_flag` which would perfectly fit the use case of overriding it once.

Reply:
Good point. I'm not quite sure where the best location for this would be. Maybe a generic `The logger shall not block, neither on logging nor on runtime exchange` under `Considerations`?